### PR TITLE
Joshua s brown/bead to unique ptr

### PR DESCRIPTION
**Pull request by JoshuaSBrown**
Memory issues part 6.

### CODE REVIEW

File: include/votca/csg/topology.h [hunk 1]
```diff
@@ -441,9 +441,9 @@ inline Bead *Topology::CreateBead(Bead::Symmetry symmetry, std::string name,
                                   std::string type, Index resnr, double m,
                                   double q) {
 
-  Bead *b = new Bead(_beads.size(), type, symmetry, name, resnr, m, q);
-  _beads.push_back(b);
-  return b;
+  _beads.emplace_back(
+      new Bead(_beads.size(), type, symmetry, name, resnr, m, q));
```
Comment (JensWehner) [inline include/votca/csg/topology.h]: You make a copy here now. Before they just copied the pointer, I think?

File: src/libcsg/modules/io/pdbwriter.cc [hunk 2]
```diff
@@ -73,7 +73,7 @@ void PDBWriter::WriteBox(const Eigen::Matrix3d &box) {
   _out << boxfrmt % a % b % c % alpha % beta % gamma;
 }
 
-void PDBWriter::writeSymmetry(Bead *bead) {
+void PDBWriter::writeSymmetry(const std::unique_ptr<Bead> &bead) {
```
Comment (JensWehner) [inline src/libcsg/modules/io/pdbwriter.cc]: I would prefer just a `const Bead& bead`

**Review by JensWehner — COMMENTED** [review]
good changes, ignore the value stuff if it complicates things.

File: include/votca/csg/topology.h (hunk 1 above)
Comment (JoshuaSBrown) [reply include/votca/csg/topology.h]: I'm confused are we looking at the same lines of code :). They were calling `new Bead` so ya they were storing the pointer in the vector but the memory was on the heap and that vector is responsible for the memory management of the bead.

File: src/libcsg/modules/io/pdbwriter.cc (hunk 2 above)
Comment (JoshuaSBrown) [reply src/libcsg/modules/io/pdbwriter.cc]: Ya but that does not work well with the vector of unique pointers stored in the topology class.

Comment (JoshuaSBrown) [reply src/libcsg/modules/io/pdbwriter.cc]: You would have to do something special to get the begin and end iterator methods to work like that. If you can figure out how I will implement it though.

File: include/votca/csg/topology.h (hunk 1 above)
Comment (JensWehner) [reply include/votca/csg/topology.h]: I did not look at the line above. YOu are right.

File: src/libcsg/modules/io/pdbwriter.cc (hunk 2 above)
Comment (JensWehner) [reply src/libcsg/modules/io/pdbwriter.cc]: you can just call `writeSymmetry(*bead)` then and dereference the pointer.

Comment (JensWehner) [reply src/libcsg/modules/io/pdbwriter.cc]: which call of this function do you think is problematic?

File: include/votca/csg/pdbwriter.h [hunk 3]
```diff
@@ -104,7 +108,7 @@ inline void PDBWriter::WriteContainer(T &container) {
   boost::format atomfrmt(
       "ATOM  %1$5d %2$-4s %3$-3s %4$1s%5$4d    %6$8.3f%7$8.3f%8$8.3f\n");
 
-  for (auto &atom : getIterable(container)) {
+  for (const auto &atom : getIterable(container)) {
```
Comment (JoshuaSBrown) [inline include/votca/csg/pdbwriter.h:106]: Jens this is the reason I pass in a unique_ptr to writeSymmetry() the beads are now stored as unique pointers in topology. When getIterable is called it is by default getting the begin and end bethods of the _beads container in topology. This means atom will be of type unique_ptr. I would have to store another container just with Bead * to do what you are asking unless I'm missing something?

Comment (JensWehner) [reply include/votca/csg/pdbwriter.h:106]: okay

**Comment by junghans** [comment]
ping @JensWehner

File: include/votca/csg/topology.h [hunk 4]
```diff
@@ -46,7 +49,9 @@ namespace csg {
 class Interaction;
 
 using MoleculeContainer = std::vector<Molecule *>;
-using BeadContainer = std::vector<Bead *>;
+// Stable vectors are necessary to be sure that the Bead pointers that point
+// to addresses in the vector are not invalidated as the vector grows.
```
Comment (JensWehner) [inline include/votca/csg/topology.h]: I mean the other way to do it is just build the vector of beads first afterwards it is stable.

File: include/votca/csg/topology.h [hunk 5]
```diff
@@ -423,7 +429,7 @@ class Topology {
 
   std::map<std::string, Index> _interaction_groups;
 
-  std::map<std::string, std::list<Interaction *> > _interactions_by_group;
+  std::map<std::string, std::list<Interaction *>> _interactions_by_group;
```
Comment (JensWehner) [inline include/votca/csg/topology.h]: can we make this 
```suggestion
  std::map<std::string, std::vector<Interaction *>> _interactions_by_group;
```

**Review by JensWehner — APPROVED** [review]
Looking good, we should make sure, that stable::vector does not kill performance, otherwise nice changes

**Comment by junghans** [comment]
@JensWehner @JoshuaSBrown:
```
/__w/csg/csg/votca/csg/src/libcsg/topology.cc:265:46: error: no viable conversion from '_Rb_tree_iterator<pair<[...], std::vector<votca::csg::Interaction *, std::allocator<votca::csg::Interaction *>>>>' to '_Rb_tree_iterator<pair<[...], std::__cxx11::list<votca::csg::Interaction *, std::allocator<votca::csg::Interaction *>>>>'
  map<string, list<Interaction *>>::iterator iter =
                                             ^
```

**Comment by junghans** [comment]
@JoshuaSBrown please resolve conflicts

**Comment by junghans** [comment]
@votca-bot format

**Comment by junghans** [comment]
@JoshuaSBrown more conflicts.

**Comment by junghans** [comment]
@votca-bot format

**Pull request merged**